File: codeforces/743div1/a.cpp
#include <bits/stdc++.h>
using namespace std;

#define pb push_back
#define eb emplace_back
#define mk make_pair
#define fi first
#define se second
#define cc(x) cout << #x << " = " << x << endl
#define ok cout << "ok" << endl
#define endl '\n'

typedef long long ll;
typedef pair<int, int> ii;
const int INF = 0x3f3f3f3f;
const double PI = acos(-1.0);

struct chapter {
  int index;
  chapter(int i) { index = i; }
};

int toposort(vector<vector<chapter>> &graph, vector<int> &degree) {
  int visited = 0;
  int ans = 1;
  queue<chapter> chapters;
  vector<int> timesNeedToRead(graph.size(), 1);

  for (int i = 0; i < degree.size(); i++) {
    if (degree[i] == 0) {
      chapters.emplace(i);
    }
  }

  while (!chapters.empty()) {
    chapter curentChapter = chapters.front();
    chapters.pop();
    visited++;
    ans = max(ans, timesNeedToRead[curentChapter.index]);

    for (chapter neighbour : graph[curentChapter.index]) {
      degree[neighbour.index]--;
      timesNeedToRead[neighbour.index] =
          max(timesNeedToRead[neighbour.index],
              timesNeedToRead[curentChapter.index] +
                  (int)(curentChapter.index > neighbour.index));
      if (degree[neighbour.index] == 0) {
        chapters.emplace(neighbour.index);
      }
    }
  }

  return (visited == graph.size() ? ans : -1);
}

int main() {
  ios_base::sync_with_stdio(false);
  int tc;
  cin >> tc;
  while (tc--) {
    int nChapters;
    cin >> nChapters;
    vector<vector<chapter>> graph(nChapters, vector<chapter>());
    vector<int> degree(nChapters, 0);
    for (int i = 0; i < nChapters; i++) {
      int kRequiredBefore;
      cin >> kRequiredBefore;
      degree[i] = kRequiredBefore;
      for (int j = 0; j < kRequiredBefore; j++) {
        int chap;
        cin >> chap;
        chap--;
        graph[chap].emplace_back(i);
      }
    }
    cout << toposort(graph, degree) << endl;
  }

  return 0;
}

File: codeforces/743div1/a.py
from typing import List
import queue


def toposort(graph: List[List[int]], degree: List[int]) -> int:
    visited = 0
    ans = 1
    queueChapters = queue.Queue()
    timesNeedToRead = [1 for _ in range(len(degree))]

    for i in range(len(degree)):
        if degree[i] == 0:
            queueChapters.put(i)
    
    while (queueChapters.empty() == False):
        curentChapter = queueChapters.get()
        visited += 1
        ans = max(ans, timesNeedToRead[curentChapter])

        for neighbour in graph[curentChapter]:
            degree[neighbour] -= 1
            timesNeedToRead[neighbour] = max(timesNeedToRead[neighbour], timesNeedToRead[curentChapter] + (curentChapter > neighbour))
            if(degree[neighbour] == 0):
                queueChapters.put(neighbour)
    
    if(visited == len(degree)):
        return ans

    return -1


def main():
    tc = int(input())
    while(tc > 0):
        tc -= 1
        nChapters = int(input())
        degree = [1 for _ in range(nChapters)]
        graph = [[] for _ in range(nChapters)]
        for i in range(nChapters):
            requiredBeforeChapters = list(map(int, input().split()))
            degree[i] = requiredBeforeChapters[0]
            requiredBeforeChapters.pop(0)
            for requiredBefore in requiredBeforeChapters:
                graph[requiredBefore-1].append(i)
        print(toposort(graph, degree))

if __name__ == "__main__":
    main()

